clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslServerAuthenticator.java

```diff
@@ -681,8 +681,8 @@ else if (connectionsMaxReauthMs == null)
                 else
                     retvalSessionLifetimeMs = zeroIfNegative(
                             Math.min(credentialExpirationMs - authenticationEndMs, connectionsMaxReauthMs));
-                if (retvalSessionLifetimeMs > 0L)
-                    sessionExpirationTimeNanos = authenticationEndNanos + 1000 * 1000 * retvalSessionLifetimeMs;
+
+                sessionExpirationTimeNanos = authenticationEndNanos + 1000 * 1000 * retvalSessionLifetimeMs;
```

Comment:
Probably worth a comment explaining that `sessionExpirationTimeNanos` should always be set to a non-null value.

Reply:
After some investigation, I found the `null` value is indicating the re-auth is disabled:
https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/network/KafkaChannel.java#L535-L543

So, maybe here, we should make sure we need re-auth (that is, we did have expiration time, if it's 0, which means it's expired, not re-auth disabled). ex:

```java
if (retvalSessionLifetimeMs > 0L || credentialExpirationMs || connectionsMaxReauthMs)
                    sessionExpirationTimeNanos = authenticationEndNanos + 1000 * 1000 * retvalSessionLifetimeMs;
```

Does that make sense?

Reply:
Adding `|| credentialExpirationMs != null || connectionsMaxReauthMs != null` to the if condition adds no value as we already know one of them has to be true to reach the if guard in general.

I think the correct fix is to delay the initialisation of `retvalSessionLifetimeMs` so that we can use a `>=` test in the expiry case and `0` in the non expiry cases while still allowing `sessionExpirationTimeNanos` being `null` to signal that re-auth is disabled.